### DISCUSSION TITLE
(void-function nil) error

### DISCUSSION THREAD

**Discussion by dr-scsi** [company-mode/company-mode#1263] [Q&A]
I'm facing a hard-to-reproduce issue with Emacs (29.0.50) where I get the error `(void-function nil)` and after that, Emacs doesn't take keyboard input (e.g. `C-x C-c`) anymore and I have to kill it from the shell.  I suspect it has something to do with company.  This is the last incarnation while I was about to insert `(def` in emacs-lisp mode.
![nil](https://user-images.githubusercontent.com/59777257/142401185-a0f1bda0-9258-4d36-b44e-22317af8ca99.png)

I have now disabled company in my init file to see if this will come back (I had disabled `company-quickhelp` but the issue came back).  But until then, have other company users seen this issue?

I use(d) only `company` and `company-quickhelp`; these are the entries in my
`.emacs`:

    (setq company-global-modes '(not latex-mode))
    (add-hook 'after-init-hook 'global-company-mode)
    (company-quickhelp-mode)

These are the versions installed from MELPA:

    company             20211112.2354
    company-quickhelp   20211115.1335

**Comment by dgutov** [comment]
I've never seen this particular error. Is the popup visible when this happens? Do you see which completion is selected?

Perhaps the variable `company-backend` gets set to nil somehow in this scenario, but it's hard to say without reproduction.

If this happened reliably enough, you could try debugging `company-fetch-metadata` to add additional reporting when `company-backend` is nil. I'm not sure how to answer the question about the circumstances, though.

The usual recommendation would be for you try "bisecting" your configuration: disable half the stuff and see whether the problem reoccurs, then repeat. But whether this is doable, depends on how often the problem repeats.

Regarding the freezing, though:

> Emacs doesn't take keyboard input (e.g. C-x C-c) anymore and I have to kill it from the shell

Have you tried pressing `C-g` multiple times? Like `C-g C-g C-g`?

**Reply by dr-scsi** [reply]
Thanks for your response.

> Is the popup visible when this happens? Do you see which completion is selected?

Popup is about to become visible, it pops up and it breaks.  I think `company` is about to collect the candidates.  As you can see in the snapshot above, Emacs chocked on the first candidate (which is `def-edebug-elem-spec` when you type `(def`)

> Perhaps the variable `company-backend` gets set to nil somehow in this scenario, ...

I'm not familiar with the internals of company, but this variable is `nil` for me when editing an .el file in Elisp mode.  Does this make sense?

> If this happened reliably enough, you could try debugging `company-fetch-metadata` to add additional reporting when `company-backend` is nil. I'm not sure how to answer the question about the circumstances, though.

This is my biggest problem: Triggering the issue.  Until now, I haven't found a recipe how to reproduce this.  At any rate, I uninstalled company for a couple of days, and the issue didn't came back.  Now I've reinstalled it from MELPA and will observe it further.  I will try to apply your hint above.

> Have you tried pressing `C-g` multiple times? Like `C-g C-g C-g`?

Yes, I did, but didn't help.

**Reply by dgutov** [reply]
> I'm not familiar with the internals of company, but this variable is nil for me when editing an .el file in Elisp mode. Does this make sense?

It's normal for it to be *usually* nil. It is assigned a value when the popup is displayed, pointing at the currently used backend.

> At any rate, I uninstalled company for a couple of days, and the issue didn't came back. Now I've reinstalled it from MELPA and will observe it further. I will try to apply your hint above.

:+1: 

> Yes, I did, but didn't help.

Odd. You can try pressing it for more times, and longer. If that doesn't help, perhaps try `kill -USR2 <emacs process pid>`, that one should land you in the debugger too.
